### ISSUE TITLE
The training time per epoch is taking too long.

### ISSUE THREAD

**Issue by LuckyDev90526** [qinliuliuqin/Medical-Detection3d-Toolkit#13]
Happy new year, @qinliuliuqin 

Thank you for sharing your work.

I am currently trying to build a medical image landmark detection model based on your solution.

I successfully trained a model using the first dataset (600 nii.gz files), and then attempted to train it with a second dataset (950 nii.gz files).

While the first dataset took approximately 2 minutes per epoch, the second dataset takes an hour per epoch.

<img width="742" height="431" alt="Image" src="https://github.com/user-attachments/assets/9bad6ddd-8599-4c06-80d4-5fde8b6d6c4b" />

I don't understand why it's taking so long.

Any help would be greatly appreciated.

Thank you.

**Comment by qinliuliuqin** [comment]
@LuckyDev90526 Thanks for your question and wish you a great new year! Could you let me know the differences between the two datasets? If 600 images took 2mins per epoch, then 950 images should only take about 3mins. Also, could you let me know why there are 432216 iterations per epoch? That number is unusually high.

**Comment by LuckyDev90526** [comment]
@qinliuliuqin , thank you for  your quick reply.

I have uploaded the sample files (CT files, mask files, and landmark label files) to Google Drive.
https://drive.google.com/drive/folders/11pQwu1ynRwlqQMuaJOg-XSUYxqw2NGEM?usp=sharing

For the first dataset, I used compressed NIfTI format (.nii.gz) CT files.
For the second dataset, I used uncompressed NIfTI format (.nii) CT files.

These files were received from a third-party vendor, so I don't know how they were generated.

I'm unsure why 432,216 iterations are needed per epoch, so I wanted to share my training configuration file with you.

I am using your latest repository, and I used the same configuration file (lmk_train_config.py) for both training runs (only the paths were changed).

Please help me.

Thank you

[lmk_train_config.py](https://github.com/user-attachments/files/24656600/lmk_train_config.py)

**Comment by LuckyDev90526** [comment]
Hi @qinliuliuqin 
I provided sample data in my last [comment](https://github.com/qinliuliuqin/Medical-Detection3d-Toolkit/issues/13#issuecomment-3757488613)
I haven't heard from you yet.
Could you please help me resolve this issue?
Thank you.
@LuckyDev90526

**Comment by qinliuliuqin** [comment]
@LuckyDev90526 Sorry, I've been tired up with other commitments. Will get back to you this weekend.

**Comment by LuckyDev90526** [comment]
Hello, @qinliuliuqin.
Could you please help me for a moment?
Thank you.

**Comment by qinliuliuqin** [comment]
Hi @LuckyDev90526,

First of all, there are 432,216 iterations because you have 864 training cases for 2001 training epochs with batch size 4. (2001 x 864 / 4 = 432216). 

Then, I noticed that in your log, the first epoch took ~1 hour, but the second epoch only took 10 mins. According to my experience, the first epoch might be ~0.5 times longer than others, but in your case it's almost 6 times, which is unusual. However, I have no idea why it's the case.

Finally, you currently set the validation interval to 1, so the validation will be conducted every epoch, which is time consuming. To speed up training, you may increase the validation interval to 100 or even higher.  This may save at least 10% of training time as you have 96 validation cases.

Let me know if you have any further questions.

**Comment by LuckyDev90526** [comment]
Hello @qinliuliuqin,
Thank you for your reply.
I understand the iterative calculation method.

The first epoch took approximately 55 minutes, and the second epoch took approximately 53 minutes (not 10 minutes).

Previously, when I trained with the first dataset, the number of iterations was 214,607 (429 x 864 / 4 = 214,607), and each epoch only took about 2 minutes.

This means that the number of iterations in the second training was twice that of the first training, but the training time took 25 times longer than the first training.

I think there might be a serious problem with the mask data generated by [mask generation script](https://github.com/qinliuliuqin/Medical-Detection3d-Toolkit/blob/master/detection3d/scripts/gen_landmark_mask.py).

I have modified the script to suit my needs.

I would appreciate it if you could review my modified mask generation script and the attached dataset file ([the second dataset](https://drive.google.com/drive/folders/15S1Q5gFDXazyVJBsyH08ychaLJEJfXUj)).

Please help me.
Thank you.

[dataset_generator.py](https://github.com/user-attachments/files/24888058/dataset_generator.py)

**Comment by qinliuliuqin** [comment]
Hi @LuckyDev90526,

Yes, you are right on the training time for the first two epochs (sorry for my misleading info). I copied the training data for 864 times to keep it consistent with your training setting. It took ~15mins for training and validation on my side. I think there must be some abnormal data in your training set which blocked the training speed. For example, some data might be very large? To debug, could you profile the training time for each batch, so we can figure out the abnormal data? 

<img width="1161" height="446" alt="Image" src="https://github.com/user-attachments/assets/0832746b-0c5f-4fca-8532-425c6a751e15" />

**Comment by LuckyDev90526** [comment]
Dear @qinliuliuqin 
Thank you for your efforts and response.
Thank you for your advice.
As you mentioned, it seems that the long training time is due to some abnormal data.
The first dataset was in a compressed format (nii.gz), but the second dataset was in an uncompressed format (nii).
Initially, since each file in the second dataset was between 900MB and 2GB, I converted them to a compressed format (nii.gz) to reduce the file size.
I am confident that the main reason for the slow training speed is due to some abnormal data in the training dataset.
Could you please provide a solution on how to filter out the abnormal data?

My PC has two GPUs. Is there a way to utilize both of them instead of just one?
In my case, I set __C.general.num_gpus to 2 in the [lmk_train_config.py](https://github.com/qinliuliuqin/Medical-Detection3d-Toolkit/blob/master/detection3d/config/lmk_train_config.py) file, but the training speed is the same as when I set it to 1.
 
Thanks

**Comment by qinliuliuqin** [comment]
Hi @LuckyDev90526

Sorry for the late response. You may profiling time costs and training file names here to identify the abnormal cases: https://github.com/qinliuliuqin/Medical-Detection3d-Toolkit/blob/0d9362a529d305b42550c4328215db48629dca72/detection3d/core/lmk_det_train.py#L253

For training with multiple GPUs, you also need to add the gpu ids. For example, training with two gpus:
```
python lmk_det_train.py --input ./config/lmk_train_config.py --gpus 0,1
```

Let me know if you have any further questions in resolving the issue.

**Comment by LuckyDev90526** [comment]
Hello @qinliuliuqin, thank you for your reply.
I checked the code snippet you provided, and it seems to be measuring the time taken per epoch (batch size).
In my case, the training time per epoch is almost constant.
My goal is to filter out certain abnormal data.
Do you have any solutions for this?

Regarding the use of multiple GPUs, I'm using the command you provided, but I'm not sure if the training speed is actually improving when using multiple GPUs (2 GPUs) compared to using a single GPU.

@qinliuliuqin, I have another serious problem.
I tried running the inference script using the intermediate weights during model training (Epoch: 472, Batch: 47436, Training Loss: 0.000104, Validation Loss: 0.000407, Time: 0.0321 seconds/volume).
However, all the landmark coordinates came out as (-1, -1, -1). This means that the intermediate model is not detecting any landmarks at all.

Could there be a problem with the training data?

I would appreciate your feedback.

Thank you.

[0-015-007_ai_annotations.nii.csv](https://github.com/user-attachments/files/25046045/0-015-007_ai_annotations.nii.csv)
[3-020-032_ai_annotations.nii.csv](https://github.com/user-attachments/files/25046046/3-020-032_ai_annotations.nii.csv)
